File: C++ Algorithms/bubble_sort.cpp
#include <iostream>
using namespace std;

/// @brief Bubble Sort Algorithm implementation in C++
/// @param arr Array of integers to be sorted
/// @param size Size of the array
void bubbleSort(int arr[], int size)
{
    while (true)
    {

        bool swapped = false;

        for (int i = 0; i < size - 1; i++)
        {
            if (arr[i] > arr[i + 1])
            {

                swapped = true;

                int temp = arr[i];
                arr[i] = arr[i + 1];
                arr[i + 1] = temp;
            }
        }

        if (!swapped)
        {
            break;
        }
    }
}

int main()
{

    int arr[] = {34, 121, 454, 22, 214, 54};
    int n = sizeof(arr) / sizeof(arr[0]);
    bubbleSort(arr, n);

    cout << "Sorted array: ";

    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }

    cout << endl;
    return 0;
}

File: java_algorithms/BubbleSort.java
import java.util.Arrays;

public class BubbleSort {

    public static void main(String[] args) {

        int[] testArray = {2, 45, 123, 5, 46, 2, 356, 7, 8, 3};
        System.out.println("Before: " + Arrays.toString(testArray));

        bubbleSort(testArray);
        System.out.println("After: " + Arrays.toString(testArray));
    }


    /**
     * Bubble Sort Algorithm implementation in Java.
     */
    public static void bubbleSort(int[] array) {

        // infinite loop , similar to while(true)
        for (; ; ) {

            boolean swapped = false;

            for (int i = 0; i < (array.length - 1); i++) {
                if (array[i] > array[i + 1]) {

                    swapped = true;

                    // traditional swap
                    int temp = array[i];
                    array[i] = array[i + 1];
                    array[i + 1] = temp;
                }
            }

            // if no swaps take place , our array has been sorted
            if (!swapped) {
                break;
            }
        }
    }
}
